#12953 Added Default Tests for Root and Homepage Pages in Wagtail Starter Projects

megatrron opened the pull request:
**#12953 Add Default Tests for Root and Homepage Pages in Wagtail Starter Projects** 
Added tests.py file in wagtail/wagtail/project_template/home/ with given code
Reinstalled wagtail locally through "pip install -e ."
Ran given tests "python manage.py test home" in created project directory.
Updated wagtail/docs/getting_started/tutorial.md

lb- commented on docs/getting_started/tutorial.md:
````diff
@@ -146,6 +146,19 @@ After the server starts, go to <http://127.0.0.1:8000> to see Wagtail’s welcom
 ```
+### Running and Writing Tests
+
+Wagtail now includes basic tests in the `home/tests.py` file when you create a new project using `wagtail start`. These tests check:
````
```suggestion
When you create a new project using `wagtail start`, there will be a set of basic tests included in the `home/tests.py` file.

These tests check:
```

Just a reword, we do not need to say 'now includes' as these docs will be read when the updates already exist. Also, we try to avoid repeating 'Wagtail' as it's the docs you are already looking at.

lb- commented on docs/getting_started/tutorial.md:
```diff
@@ -146,6 +146,19 @@ After the server starts, go to <http://127.0.0.1:8000> to see Wagtail’s welcom
+
+Wagtail now includes basic tests in the `home/tests.py` file when you create a new project using `wagtail start`. These tests check:
+
+- If the **Root Page** (ID=1) is automatically created.
+- If the **Homepage** is created as a child of the root page.
+- If a **BlogIndexPage** (or other page types) can be added.
```
Just a refinement, avoiding bold and using the model as a code formatted proper noun.

```suggestion
-   If the _root `Page`_ (ID=1) is automatically created.
-   If the _home `Page`_ is created as a child of the root page.
-   If a `BlogIndexPage` (or other page types) can be added.
```

Also, adding extra spaces to the list, while we do not use Prettier in MD docs, it's going to get these extra spaces when we do manual clean up sometimes.

lb- commented on docs/getting_started/tutorial.md:
```diff
@@ -146,6 +146,19 @@ After the server starts, go to <http://127.0.0.1:8000> to see Wagtail’s welcom
+- If a **BlogIndexPage** (or other page types) can be added.
+
+#### Running the Tests
+To run the tests, navigate to your project folder and run:
```
```suggestion

To run the tests, navigate to your project folder and run:
```

Extra line break.

lb- commented on docs/getting_started/tutorial.md:
````diff
@@ -146,6 +146,19 @@ After the server starts, go to <http://127.0.0.1:8000> to see Wagtail’s welcom
+To run the tests, navigate to your project folder and run:
+
+```sh
+python manage.py test home
````
````suggestion
python manage.py test home
```
````

Missed the closing backticks.

lb- commented on wagtail/project_template/home/tests.py:
```diff
@@ -0,0 +1,39 @@
+        """
+        Tests if BlogIndexPage can be created.
+        """
+        raise NotImplementedError("The tests are not implemented yet.")
```
Line break needed at the end of the file, minor formatting but it just means the CI will not complain.
Best to run your code formatter for new PRs.

See https://docs.wagtail.org/en/stable/contributing/python_guidelines.html#pep8

lb- commented on wagtail/project_template/home/tests.py:
```diff
@@ -0,0 +1,39 @@
+# home/tests.py
```
```suggestion
```

Do we need this comment? it does not look like we do this in any other template files.

lb- commented on wagtail/project_template/home/tests.py:
```diff
@@ -0,0 +1,39 @@
+# home/tests.py
+
+from django.test import TestCase
```
I think we should try to use our documented testing approach in https://docs.wagtail.org/en/stable/advanced_topics/testing.html


```suggestion
from wagtail.test.utils import WagtailPageTestCase
```

Also, have a review of that page for some other tips/practice we recommend in our docs.

lb- commented on wagtail/project_template/home/tests.py:
```diff
@@ -0,0 +1,39 @@
+    def test_homepage_create(self):
+        root_page = Page.objects.get(pk=1)
+        self.homepage = HomePage(title='Home')
+        root_page.add_child(instance=self.homepage)
```
What is this actually testing?

Do we need some assertions here?

Same for the test above.

lb- reviewed (changes requested):
Thanks @megatrron  - great start here.

I have some specific feedback for the documentation.

As for the unit tests, I know the PR had the code that you are using but I would like us to go a touch deeper and actually have a few assertions in the unit tests.

I see the goal here as two fold

1. A starting test file that actually tests a few things that are included within the template
2. A test file that can be expanded on 

I think we should try to add a few simple assertions to the existing tests and if possible use something that comes with our documented testing class. https://docs.wagtail.org/en/stable/advanced_topics/testing.html

megatrron commented:
I will work on this. Thanks for reviewing @lb-

megatrron commented:
@lb- I updated the documentation as you said and added assertions in tests .

gasman commented on docs/getting_started/tutorial.md:
````diff
@@ -146,6 +146,23 @@ After the server starts, go to <http://127.0.0.1:8000> to see Wagtail’s welcom
+```sh
+python manage.py test home
+```
 
````
As a general point - we need to think very carefully about any changes to the tutorial, particularly ones so early in the steps, as this will often be the first impression that new users receive of Wagtail.

* We want to ensure that users following the tutorial see tangible results as quickly as possible. It's good to get users into the mindset of testing early, but do we really want to take them on a tangent before they've even seen the admin interface?
* The description of what these tests are checking includes concepts that haven't been explained yet - such as `BlogIndexPage` (is this actually true any more?) and, particularly, the idea that "root" is a page in its own right - I'd actually consider this to be an internal implementation detail that new users shouldn't have to know, and we'd generally avoid mentioning it directly, through wording like "create a page at the root level" rather than "create a page as a child of the root page".
* Have we checked that these tests continue to work correctly after we've extended HomePage?

lb- replied:
@megatrron can you move the content to the project template page in the docs instead?

Probably under the Django settings section, before the Docker file one.

megatrron commented:
@lb- done. If more modifications required, please let me know.

lb- commented:
Epic. Thanks @megatrron - I've started the CI and flagged this as ready for another review.

megatrron commented:
@lb- 2 checks failed , do i need to modify test.yml file or anything else or its fine. I am new to this so please guide me. Thank you

lb- commented on docs/reference/project_template.md:
````diff
@@ -97,6 +98,24 @@ The Django settings files are split up into `base.py`, `dev.py`, `production.py`
 On production servers, we recommend that you only store secrets in ``local.py`` (such as API keys and passwords). This can save you headaches in the future if you are ever trying to debug why a server is behaving badly. If you are using multiple servers which need different settings then we recommend that you create a different ``production.py`` file for each one.
 ```
 
+### Running and Writing Tests
````
```suggestion
### Writing and running tests
```

Titles can just be sentence case as per the current convention in the docs, let's switch the words around also.

lb- commented on docs/reference/project_template.md:
```diff
@@ -97,6 +98,24 @@ The Django settings files are split up into `base.py`, `dev.py`, `production.py`
+-   If the _home `Page`_ is created as a child of the root page.
+-   If a `BlogIndexPage` (or other page types) can be added.
+
+#### Running the Tests
```
```suggestion
#### Running the tests
```

lb- commented on docs/reference/project_template.md:
```diff
@@ -97,6 +98,24 @@ The Django settings files are split up into `base.py`, `dev.py`, `production.py`
+These tests check:
+
+-   If the _root `Page`_ (ID=1) is automatically created.
+-   If the _home `Page`_ is created as a child of the root page.
```
```suggestion
-   That the _home `Page`_ can be created as a child of the root page.
```

lb- commented on docs/reference/project_template.md:
```diff
@@ -97,6 +98,24 @@ The Django settings files are split up into `base.py`, `dev.py`, `production.py`
+
+-   If the _root `Page`_ (ID=1) is automatically created.
+-   If the _home `Page`_ is created as a child of the root page.
+-   If a `BlogIndexPage` (or other page types) can be added.
```
```suggestion
-   A placeholder test for other sub-pages, such as the `BlogIndexPage`, ready for implementation.
```

This test is not actually implemented at the start so we should be clear about this.

**Update** - maybe we just remove this placeholder test, which just fails, to avoid confusion.

lb- commented on wagtail/project_template/home/tests.py:
```diff
@@ -0,0 +1,40 @@
+        root_page = Page.objects.get(pk=1)
+        self.homepage = HomePage(title='Home')
+        root_page.add_child(instance=self.homepage)
+        
```
Note, AI suggestion, but maybe we should do something else here to actually have a basic starting set of tests.

```suggestion
    def test_homepage_status_code(self):
        response = self.client.get(reverse('home'))
        self.assertEqual(response.status_code, 200)

    def test_homepage_template_used(self):
        response = self.client.get(reverse('home'))
        self.assertTemplateUsed(response, 'home/index.html')
```

Be sure to run this code yourself, ensuring the tests run for a new project.

lb- commented on wagtail/project_template/home/tests.py:
```diff
@@ -0,0 +1,40 @@
+        self.homepage = HomePage(title='Home')
+        root_page.add_child(instance=self.homepage)
+        
+    def test_your_test(self):
+        """
+        Tests if BlogIndexPage can be created.
+        """
+
+        raise NotImplementedError("The tests are not implemented yet.")
```
I think we should choose to start with a set of tests that pass. It's possible it could cause confusion if the tutorial starting point does not pass.

I would recommend we actually remove this 'placeholder' test and just add a few more starting tests.

lb- commented on docs/reference/project_template.md:
````diff
@@ -97,6 +98,24 @@ The Django settings files are split up into `base.py`, `dev.py`, `production.py`
+To run the tests, navigate to your project folder and run:
+
+```sh
+python manage.py test home
````
```suggestion
python manage.py test
```

Would we not just recommend the generic test command? New developers may get confused if they have a different app name or move this later and it's not in the `home` module.

lb- commented on docs/reference/project_template.md:
````diff
@@ -97,6 +98,24 @@ The Django settings files are split up into `base.py`, `dev.py`, `production.py`
+```sh
+python manage.py test home
+```
+
````
```suggestion

For more about testing within Django see ..., or Wagtail specific unit testing utils, see ...
```

Let's add cross links to the Django docs & the Wagtail docs.
https://docs.djangoproject.com/en/5.1/topics/testing/overview/

Be sure to use the correct linking mechanism in the docs, not just raw URLs. See https://docs.wagtail.org/en/stable/contributing/documentation_guidelines.html#links

lb- reviewed (changes requested):
Thanks @megatrron - this is on the way, I think it needs a bit more work until we can fully review & get this merged.

I have some suggestions, I think we can do a bit better with having a few more tests but also avoid the placeholder tests.  Unless you want to maybe add the `@skip` decorator for that one?

e.g. `@skip("Example test...")`

Thanks for pursuing this, super keen to get this in, just needs a bit more polish.

I have not tested this locally yet, will do on the next round of reviews.

megatrron commented:
@lb- apologies for so much delay as my end semester exams were going on.
I agree with your opinion on removing placeholder test. I have added the tests as suggested with a bit of modifications. I have ran the tests after adding necessary views with url mapping on a fresh project and they ran successfully.
![2025-05-09_18-06](https://github.com/user-attachments/assets/1d84e7cf-705f-4a47-b8f4-9e75507d6ef8)
I have also updated the documentation as per your instructions.
Please point out if there is any mistake from my side or any further improvents are needed.
Thanks

lb- reviewed (approved):
Looking good, I will make some wording tweaks to the test docstrings when merging.

Thank you @megatrron  for working through the feedback and getting this through.

It's a great solid start for testing in our project template.
